Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Sharing/MarkerDetectorSpatialLocalizer.cs

```diff
@@ -30,47 +37,106 @@ private void OnValidate()
             if (markerDetector == null)
             {
                 Debug.LogWarning("Marker detector not appropriately set for MarkerDetectorSpatialLocalizer");
             }
+        }
+
+        public override bool TryDeserializeSettings(BinaryReader reader, out MarkerDetectorLocalizationSettings settings)
+        {
+            return MarkerDetectorLocalizationSettings.TryDeserialize(reader, out settings);
+        }
 
-            spatialCoordinateService = new MarkerDetectorCoordinateService(markerDetector, debugLogging);
+        public override ISpatialLocalizationSession CreateLocalizationSession(MarkerDetectorLocalizationSettings settings)
+        {
+            return new LocalizationSession(this, settings);
         }
 
-        /// <inheritdoc/>
-        protected override async Task<ISpatialCoordinate> GetHostCoordinateAsync(Guid token)
+        private class SpatialCoordinate : TransformSpatialCoordinate<int>
         {
-            DebugLog("Getting host coordinate", token);
+            private Marker marker;
+
+            public Marker Marker
+            {
+                get => marker;
+                set
+                {
+                    if (value != null && value.Id != Id)
+                    {
+                        throw new InvalidOperationException("Setting a marker with the wrong id.");
+                    }
+
+                    if (marker != value)
+                    {
+                        marker = value;
+                        gameObject.transform.position = marker.Position;
+                        gameObject.transform.rotation = marker.Rotation;
+                    }
+                }
+            }
+
+            internal GameObject GameObject => gameObject;
 
-            using (CancellationTokenSource cts = new CancellationTokenSource())
+            public override LocatedState State => marker == null ? base.State : LocatedState.Tracking;
+
+            public SpatialCoordinate(int id)
+                : base(id)
             {
+            }
+
+            public SpatialCoordinate(Marker marker)
+                : this(marker?.Id ?? throw new ArgumentNullException(nameof(marker)))
+            {
+                Marker = marker;
+            }
+        }
+
+        private class LocalizationSession : SpatialLocalizationSession<MarkerDetectorLocalizationSettings>
+        {
+            private readonly MarkerDetectorSpatialLocalizer localizer;
+
+            public LocalizationSession(MarkerDetectorSpatialLocalizer localizer, MarkerDetectorLocalizationSettings settings)
+                : base(settings)
+            {
+                this.localizer = localizer;
+            }
+
+            public override async Task<ISpatialCoordinate> LocalizeAsync(CancellationToken cancellationToken)
+            {
+                localizer.DebugLog("Getting host coordinate");
+                localizer.markerDetector.SetMarkerSize(Settings.MarkerSize);
+
```

Comment:
it feels like this should be calling the SpatialCoordinateService compared to directly using the marker detector.

Reply:
But this organization does raise the question of whether a SpatialLocalizer can support multiple sessions at the same time. Having multiple sessions seems beyond the scope of what we should test for getting things to initially work.

---
In reply to: [292199660](https://github.com/microsoft/MixedRealityToolkit-Unity/pull/4810#discussion_r292199660) [](ancestors = 292199660)

Reply:
Yes, this should be going through the SpatialCoordinateService, I had slashed and burned to get things working. I'll be putting this back the way it was, with refactors necessary to actually configure the marker-specific version before localizing